engine/src/main/java/org/hibernate/validator/internal/engine/ValidatorImpl.java

```diff
@@ -533,13 +538,47 @@ private boolean validateConstraint(ValidationContext<?> validationContext,
 			if ( !propertyPathComplete ) {
 				valueContext.appendNode( propertyMetaData );
 			}
-			// set the unwrapping mode for this validation
-			valueContext.setUnwrapMode( propertyMetaData.unwrapMode() );
+
+			if ( TYPE_USE.equals( metaConstraint.getElementType().name() ) ) {
+				// TYPE_USE constraints always require UNWRAP
+				valueContext.setUnwrapMode( UnwrapMode.UNWRAP );
+			}
+			else {
+				// set the unwrapping mode for this validation
+				valueContext.setUnwrapMode( propertyMetaData.unwrapMode() );
+			}
 		}
 		else {
 			valueContext.appendBeanNode();
 		}
 
+		validationSuccessful = validateMetaConstraint( validationContext, valueContext, metaConstraint );
+
+		// reset the unwrapping mode
+		valueContext.setUnwrapMode( UnwrapMode.AUTOMATIC );
+		return validationSuccessful;
+	}
+
+	private boolean validateTypeConstraint(ValidationContext<?> validationContext, ValueContext<?, Object> valueContext, boolean propertyPathComplete,
```

Comment:
Hmm , good question. Obviously the executable validation is not fully unified with the other validate methods. There is also `OptionalTypeAnnotationConstraintOnMethodTest` which tests parameter validation. 

Return value validation could need another test ...